pkg/brokerapi/openservicebroker/open_service_broker_client.go

```diff
@@ -78,59 +116,78 @@ func (c *openServiceBrokerClient) GetCatalog() (*brokerapi.Catalog, error) {
 	return &catalog, nil
 }
 
-func (c *openServiceBrokerClient) CreateServiceInstance(ID string, req *brokerapi.ServiceInstanceRequest) (*brokerapi.ServiceInstance, error) {
-	jsonBytes, err := json.Marshal(req)
-	if err != nil {
-		return nil, err
-	}
-
+func (c *openServiceBrokerClient) CreateServiceInstance(ID string, req *brokerapi.CreateServiceInstanceRequest) (*brokerapi.CreateServiceInstanceResponse, error) {
 	url := fmt.Sprintf(serviceInstanceFormatString, c.broker.Spec.URL, ID)
-
 	// TODO: Handle the auth
-	createHTTPReq, err := http.NewRequest("PUT", url, bytes.NewReader(jsonBytes))
+	resp, err := util.SendRequest(c.client, http.MethodPut, url, req)
 	if err != nil {
-		return nil, err
+		return nil, errRequest{message: err.Error()}
 	}
+	defer resp.Body.Close()
 
-	glog.Infof("Doing a request to: %s", url)
-	resp, err := c.client.Do(createHTTPReq)
-	if err != nil {
-		return nil, err
+	createServiceInstanceResponse := brokerapi.CreateServiceInstanceResponse{}
+	if err := util.ResponseBodyToObject(resp, &createServiceInstanceResponse); err != nil {
+		return nil, errResponse{message: err.Error()}
 	}
-	defer resp.Body.Close()
 
-	// TODO: Align this with the actual response model.
-	si := brokerapi.ServiceInstance{}
-	if err := util.ResponseBodyToObject(resp, &si); err != nil {
-		return nil, err
+	switch resp.StatusCode {
+	case http.StatusCreated:
+		return &createServiceInstanceResponse, nil
+	case http.StatusOK:
+		return &createServiceInstanceResponse, nil
+	case http.StatusAccepted:
+		glog.Infof("Asynchronous response received. Polling broker.")
+		if err := c.pollBroker(ID, createServiceInstanceResponse.Operation); err != nil {
```

Comment:
I think this is a fine first start.  Note that from the calling standpoint, an asynchronous provision just looks like a long blocking call.

I think that to be most useful for clients of this API, we need to provide feedback to the caller immediately that the request was accepted.  In our case, this will allow the controller to update the status of the Instance right away.

So, I think that long-term the right API is one where you can pass a callback to be executed on completion of the provision request.  If you get back a 200, you can execute it immediately.  If you get back a 202, the client should schedule a poll, possibly retrying, until success or failure is determined, at which point it would execute the passed in callback and pass that callback the result of the operation.

Does that make sense?